File: scala/bob/src/main/scala/Bob.scala
object Bob {
  trait Teenager {
    def blanketStatment: String  = "Whatever."
    def forcefulQuestion: String = "Calm down, I know what I'm doing!"
    def question: String         = "Sure."
    def shouting: String         = "Whoa, chill out!"
    def silence: String          = "Fine. Be that way!"
  }

  case object Bob extends Teenager

  def response(statement: String): String = {
    statement.trim match {
      case s if s.isEmpty                     => Bob.silence
      case s if shouting(s) && questioning(s) => Bob.forcefulQuestion
      case s if shouting(s)                   => Bob.shouting
      case s if questioning(s)                => Bob.question
      case _                                  => Bob.blanketStatment
    }
  }

  def shouting(statement: String): Boolean = {
    statement.forall(!_.isLower) && statement.exists(_.isLetter)
  }

  def questioning(statement: String): Boolean =
    statement.endsWith("?")
}

File: typescript/bob/bob.ts
enum TeenagerResponse {
  BlanketStatement = 'Whatever.',
  ForcefulQuestion = 'Calm down, I know what I\'m doing!',
  Question = 'Sure.',
  Silence = 'Fine. Be that way!',
  Yelling = 'Whoa, chill out!',
}

type DetermineStatement = (statement: string) => boolean;

const isShouting: DetermineStatement = (statement) => (
  (statement === statement.toUpperCase()) && /[a-zA-Z]/.test(statement)
);

const isQuestion: DetermineStatement = (statement) => statement.endsWith('?');

export const speakToBob = (statement: string): TeenagerResponse => {
  const trimmed = statement.trim();

  if (!trimmed) {
    return TeenagerResponse.Silence;
  } else if (isShouting(trimmed) && isQuestion(trimmed)) {
    return TeenagerResponse.ForcefulQuestion;
  } else if (isShouting(trimmed)) {
    return TeenagerResponse.Yelling;
  } else if (isQuestion(trimmed)) {
    return TeenagerResponse.Question;
  }

  return TeenagerResponse.BlanketStatement;
};
